File: src/main/java/com/dude/dms/backend/service/HistoricalCrudService.java
package com.dude.dms.backend.service;

import com.dude.dms.backend.data.DataEntity;
import com.dude.dms.backend.data.Diffable;
import com.dude.dms.backend.data.Historical;
import com.dude.dms.backend.data.history.History;
import org.springframework.beans.factory.annotation.Autowired;

public abstract class HistoricalCrudService<T extends DataEntity & Historical<U> & Diffable<T>, U extends History> extends CrudService<T> {

    @Autowired
    private HistoryCrudService<T, U> historyService;

    protected abstract U createHistory(T entity, String text, boolean created, boolean edited, boolean deleted);

    @Override
    public T save(T entity) {
        T before = load(entity.getId());
        T after = super.save(entity);
        historyService.create(createHistory(after, before.diff(after), false, true, false));
        return after;
    }

    @Override
    public T create(T entity) {
        T created = super.create(entity);
        historyService.create(createHistory(created, null, true, false, false));
        return created;
    }

    @Override
    public void delete(T entity) {
        super.delete(entity);
        historyService.create(createHistory(entity, null, false, false, true));
    }
}

File: src/main/kotlin/com/dude/dms/backend/service/HistoricalCrudService.kt
package com.dude.dms.backend.service

import com.dude.dms.backend.data.DataEntity
import com.dude.dms.backend.data.Diffable
import com.dude.dms.backend.data.Historical
import com.dude.dms.backend.data.history.History
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.jpa.repository.JpaRepository

abstract class HistoricalCrudService<T, U : History>(jpaRepository: JpaRepository<T, Long>) : CrudService<T>(jpaRepository) where T : DataEntity, T : Historical<U>, T : Diffable<T> {

    @Autowired
    private lateinit var historyService: HistoryCrudService<T, U>

    protected abstract fun createHistory(entity: T, text: String?, created: Boolean, edited: Boolean, deleted: Boolean): U

    override fun save(entity: T): T {
        val before = load(entity.id)
        val after = super.save(entity)
        historyService.create(createHistory(after, before.diff(after), created = false, edited = true, deleted = false))
        return after
    }

    override fun create(entity: T): T {
        val created = super.create(entity)
        historyService.create(createHistory(created, null, created = true, edited = false, deleted = false))
        return created
    }

    override fun delete(entity: T) {
        super.delete(entity)
        historyService.create(createHistory(entity, null, created = false, edited = false, deleted = true))
    }
}